pkg/worker/grpc_worker.go

```diff
@@ -0,0 +1,285 @@
+		if v != nil {
+			headers[k] = &pb.HeaderValue{
+				Value: v,
+			}
+			if len(v) > 0 {
+				headersOld[k] = v[0]
+			}
+		}
+	}
+
+	query := make(map[string]*pb.QueryValue)
+	queryOld := make(map[string]string)
+	for k, v := range trigger.Query {
+		if v != nil {
+			query[k] = &pb.QueryValue{
+				Value: v,
+			}
+			if len(v) > 0 {
+				queryOld[k] = v[0]
+			}
+		}
+	}
+
+	triggerRequest := &pb.TriggerRequest{
+		Data:     trigger.Body,
+		MimeType: mimeType,
+		Context: &pb.TriggerRequest_Http{
+			Http: &pb.HttpTriggerContext{
+				Path:           trigger.Path,
+				Method:         trigger.Method,
+				QueryParams:    query,
+				QueryParamsOld: queryOld,
+				Headers:        headers,
+				HeadersOld:     headersOld,
+				PathParams:     trigger.Params,
+			},
+		},
+	}
+
+	// construct the message
+	message := &pb.ServerMessage{
+		Id: ID,
+		Content: &pb.ServerMessage_TriggerRequest{
+			TriggerRequest: triggerRequest,
+		},
+	}
+
+	// send the message
+	err := s.send(message)
+
+	if err != nil {
+		// There was an error enqueuing the message
+		return nil, err
+	}
+
+	// wait for the response
+	triggerResponse := <-returnChan
+
+	httpResponse := triggerResponse.GetHttp()
+
+	if httpResponse == nil {
+		return nil, fmt.Errorf("fatal: Error handling event, incorrect response received from function")
+	}
+
+	fasthttpHeader := &fasthttp.ResponseHeader{}
+
+	for key, val := range httpResponse.GetHeaders() {
+		headerList := val.Value
+		if key == "Set-Cookie" || key == "Cookie" {
+			for _, v := range headerList {
+				fasthttpHeader.Add(key, v)
+			}
+		} else if len(headerList) > 0 {
+			fasthttpHeader.Set(key, headerList[0])
+		}
+	}
+
+	response := &triggers.HttpResponse{
+		Body: triggerResponse.Data,
+		// No need to worry about integer truncation
+		// as this should be a HTTP status code...
+		StatusCode: int(httpResponse.Status),
+		Header:     fasthttpHeader,
+	}
+
+	return response, nil
+}
+
+func (s *grpcWorkerBase) HandleEvent(trigger *triggers.Event) error {
+	// Generate an ID here
+	ID, returnChan := s.newTicket()
+	triggerRequest := &pb.TriggerRequest{
+		Data:     trigger.Payload,
+		MimeType: http.DetectContentType(trigger.Payload),
+		Context: &pb.TriggerRequest_Topic{
+			Topic: &pb.TopicTriggerContext{
+				Topic: trigger.Topic,
+				// FIXME: Add missing fields here...
+			},
+		},
+	}
+
+	// construct the message
+	message := &pb.ServerMessage{
+		Id: ID,
+		Content: &pb.ServerMessage_TriggerRequest{
+			TriggerRequest: triggerRequest,
+		},
+	}
+
+	// send the message
+	err := s.send(message)
+
+	if err != nil {
+		// There was an error enqueuing the message
+		return err
+	}
+
+	// wait for the response
+	// FIXME: Need to handle timeouts here...
+	response := <-returnChan
+
+	topic := response.GetTopic()
+
+	if topic == nil {
+		// Fatal error in this case
+		// We don't have the correct response type for this handler
+		return fmt.Errorf("Fatal: Error handling event, incorrect response received from function")
```

Comment:
should probably create a custom FatalError type ..

Reply:
Have put in a new errors package and add in a new FatalError type for now (can add additional types as we refactor).